utils/merkletrie/fsnoder/file.go

```diff
@@ -0,0 +1,73 @@
+package fsnoder
+
+import (
+	"bytes"
+	"fmt"
+	"hash/fnv"
+
+	"gopkg.in/src-d/go-git.v4/utils/merkletrie/noder"
+)
+
+// file values represent file-like noders in a merkle trie.
+type file struct {
+	name     string // relative
+	contents string
+	hash     []byte // memoized
+}
+
+// newFile returns a noder representing a file with the given contents.
+func newFile(name, contents string) (*file, error) {
+	if name == "" {
+		return nil, fmt.Errorf("files cannot have empty names")
+	}
+
+	return &file{
+		name:     name,
+		contents: contents,
+	}, nil
+}
+
+// The hash of a file is just its contents.
+// Empty files will have the fnv64 basis offset as its hash.
+func (f *file) Hash() []byte {
+	if f.hash == nil {
+		h := fnv.New64a()
+		h.Write([]byte(f.contents))
+		f.hash = []byte{}
+		f.hash = h.Sum(f.hash)
+	}
+
+	return f.hash
+}
+
+func (f *file) Name() string {
+	return f.name
+}
+
+func (f *file) IsDir() bool {
+	return false
+}
+
+func (f *file) Children() ([]noder.Noder, error) {
+	return noder.NoChildren, nil
```

Comment:
We need a way for `Children()` to return "there are no childrens": We can use `nil` or `[]noder.Noder{}`.

As `reflect.DeepEqual(nil, []noder.Noder{})` is `false`, we should not mix both of them, so we have to use one or the other.

Nil is not a good options as it is difficult to compare children from one noder with children with the other when one of them is nil.

So the best way to mean "there is no children" is to use an empty slice.

Now, the best way to communicate that we have chosen an empty slice over nil is to define a public var for it. The definition of `noder.NoChildren` explains that:

```
// NoChildren represents the children of a noder without children.
var NoChildren = []Noder{}
```

And it is also explained at the `noder.Noder` interface definition of `Children()`:

```
// Children returns the children of the element.  Note that empty
// directory-like noders and file-like noders will both return
// NoChildren.
        Children() ([]Noder, error) 
```

I would galdly change the name of the var for something better, but nothing comes to my mind right now.